File: 15_Semantic_Rules/ets_cases/15.5_Invariance_Covariance_Contravariance/cross_lang_verify/SEM_15_05_Variance.java
/**
 * Java cross-language verification for 15.5 Invariance, Covariance, Contravariance
 *
 * Key differences:
 * - Java supports use-site variance (? extends / ? super), ArkTS does not
 * - Swift supports declaration-site variance (inout), ArkTS does not
 * - All three support return type covariance and parameter contravariance
 */
class SEM_15_05_Variance {
    static int pass = 0;
    static int fail = 0;

    static void check(boolean cond, String msg) {
        if (cond) { pass++; System.out.println("  PASS " + msg); }
        else { fail++; System.out.println("  FAIL " + msg); }
    }

    // Generic invariance demonstration
    static class Box<T> {
        T value;
        Box(T v) { this.value = v; }
        T get() { return value; }
        void set(T v) { this.value = v; }
    }

    // Return type covariance
    static class Animal {
        Animal reproduce() { return new Animal(); }
    }
    static class Dog extends Animal {
        @Override
        Dog reproduce() { return new Dog(); }  // covariant return type
    }

    // Parameter contravariance via functional interface
    @FunctionalInterface
    interface Consumer<T> {
        void accept(T t);
    }

    public static void main(String[] args) {
        // === Generic invariance ===
        // ArkTS: let b: Box<Dog> = new Box<Dog>(new Dog()); let a: Box<Animal> = b; // ERROR
        // Java:  same - generic invariance
        {
            Box<Dog> dogBox = new Box<>(new Dog());
            // Box<Animal> animalBox = dogBox;  // COMPILE ERROR (invariance)
            check(true, "generic invariance (Java: same as ArkTS)");
        }

        // === Return type covariance ===
        // ArkTS: override reproduce(): Dog { ... }  // covariant return OK
        // Java:  @Override Dog reproduce() { ... }  // covariant return OK (Java 5+)
        {
            Dog d = new Dog();
            Animal result = d.reproduce();
            check(result instanceof Dog, "return type covariance (Java: same as ArkTS)");
        }

        // === Parameter contravariance via Consumer ===
        // ArkTS: function type (Animal) -> void can accept (Dog) -> void
        // Java:  Consumer<Animal> can accept Dog (wildcard)
        {
            Consumer<Dog> dogConsumer = (Dog d) -> System.out.println("  INFO dog");
            // Consumer<Animal> animalConsumer = dogConsumer;  // Java: invariance
            // Java uses wildcards: Consumer<? super Dog>
            Consumer<? super Dog> superConsumer = (Animal a) -> System.out.println("  INFO animal");
            superConsumer.accept(new Dog());
            check(true, "parameter contravariance (Java: wildcard ? super)");
        }

        // === Java-specific: use-site variance (? extends) ===
        // ArkTS does NOT support use-site variance
        {
            Box<Dog> dogBox = new Box<>(new Dog());
            // Java wildcard: allows read-only covariance at use site
            Box<? extends Animal> animalBox = dogBox;
            Animal a = animalBox.get();
            check(a instanceof Dog, "use-site variance (? extends) - Java specific, ArkTS does not support");
        }

        // === Java-specific: use-site variance (? super) ===
        {
            Box<Animal> animalBox = new Box<>(new Animal());
            Box<? super Dog> dogBox = animalBox;
            dogBox.set(new Dog());
            check(true, "use-site variance (? super) - Java specific, ArkTS does not support");
        }

        System.out.println("\n=== Summary ===");
        System.out.println("Pass: " + pass + "  Fail: " + fail);
        if (fail > 0) { System.exit(1); }
        System.out.println("verified");
    }
}

File: 15_Semantic_Rules/ets_cases/15.5_Invariance_Covariance_Contravariance/cross_lang_verify/SEM_15_05_Variance.swift
/**
 * Swift cross-language verification for 15.5 Invariance, Covariance, Contravariance
 *
 * Key difference: Swift supports declaration-site variance (generic where clauses),
 * ArkTS and Java do not.
 */
import Foundation

var pass = 0
var fail = 0

func check(_ cond: Bool, _ msg: String) {
    if cond { pass += 1; print("  PASS \(msg)") }
    else { fail += 1; print("  FAIL \(msg)") }
}

// Generic invariance
class Box<T> {
    var value: T
    init(_ v: T) { self.value = v }
    func get() -> T { return value }
    func set(_ v: T) { self.value = v }
}

// Return type covariance
class Animal {
    func reproduce() -> Animal { return Animal() }
    func speak() -> String { return "..." }
}
class Dog: Animal {
    override func reproduce() -> Dog { return Dog() }  // covariant return
    override func speak() -> String { return "Woof!" }
}

// Parameter contravariance via closure
typealias Consumer<T> = (T) -> Void

// === Generic invariance ===
// Swift: same as ArkTS/Java - generics are invariant by default
do {
    let dogBox = Box(Dog())
    // let animalBox: Box<Animal> = dogBox  // COMPILE ERROR (invariance)
    check(true, "generic invariance (Swift: same as ArkTS/Java)")
}

// === Return type covariance ===
// Swift supports covariant return types via override
do {
    let d = Dog()
    let result: Animal = d.reproduce()
    check(result is Dog, "return type covariance (Swift: same as ArkTS)")
}

// === Parameter contravariance ===
// Swift closures support contravariance naturally
do {
    let dogConsumer: Consumer<Dog> = { d in print("  INFO dog") }
    // Swift: Consumer<Animal> is not subtype of Consumer<Dog> (invariance)
    // But function types are contravariant in parameters
    let animalFunc: (Animal) -> Void = { a in print("  INFO animal") }
    let dogFunc: (Dog) -> Void = animalFunc  // ✅ function type parameter contravariance
    dogFunc(Dog())
    check(true, "parameter contravariance (Swift: function type parameter contravariance)")
}

// === Swift-specific: declaration-site variance ===
// ArkTS and Java do NOT support declaration-site variance
// Swift supports generic where clauses
do {
    func process<T: Animal>(_ t: T) -> String {
        return t.speak()
    }
    let result = process(Dog())
    check(result == "Woof!", "declaration-site variance (Swift specific, ArkTS does not support)")
}

print("\n=== Summary ===")
print("Pass: \(pass)  Fail: \(fail)")
if fail > 0 { exit(1) }
print("verified")

extension Dog {
    func bark() -> String { return "Woof!" }
}
